### ISSUE TITLE
Provide functionality to reload library namespaces

### ISSUE THREAD

**Issue by jpmonettas** [tonsky/clj-reload#10]
First thanks for the great library, I've replaced tools.namespace with it and I'm much happier.

Maybe this is outside the scope of this project, but given all the nice functionality to reload namespaces in topological order I think it would be great to have a function like `clj-reload.core/reload` to reload namespaces by regex (like with :only) but that works on the entire classpath and which doesn't take into account or modifies the current reload state.

This can be useful in situations where you go and re-eval libraries functions (like to add debug statements), so after you are done you can do something like `(clj-reload.core/reload-all  #"ring.*")`, and you don't need to think what functions you changed nor restart the repl.

**Comment by tonsky** [comment]
Wait in that example where is ring? In jar? How do you modify it?

**Comment by jpmonettas** [comment]
Yeah, in a jar. So in Cider for example you can jump to any code, doesn't matter if it is in a library inside a jar, then modify it and re-evaluate it to try something (you don't save the file, since it is in a jar). I normally use like `(require 'lib.core :reload-all)` to bring stuff back, but there is no way to give it a top namespace or a regex.

**Comment by tonsky** [comment]
And as you modify it, it will rewrite the jar file on disk?

**Comment by jpmonettas** [comment]
This functionality is coming from Emacs ability to open buffers showing files inside jars, but currently it can't change it, so no. Which makes it useful because you can change library code as you wish, for trying stuff, and you don't risk breaking anything permanently.

**Comment by tonsky** [comment]
How will reload then reload it? I am confused

**Comment by jpmonettas** [comment]
Reload would help to bring everything to the original state without having to restart the repl.

**Comment by velios** [comment]
Greetings to all. Sorry to intrude, perhaps it would be appropriate to mention that such functionality could be useful for solve task like this [issue](https://github.com/flow-storm/flow-storm-debugger/issues/171). Perhaps this will allow better explain why functionality like that might be needed at all.

**Comment by jpmonettas** [comment]
Yeah, thanks @velios , I should have mentioned that also, I was trying to give a more general description of why I think this could be a cool feature outside a specific tool like FlowStorm.

**Comment by tonsky** [comment]
Okay I see the use-case. I have a few questions though, as this seems asymmetrical to me yet:

- If you can change library code and reload it manually to see changes applied
- Obv. clj-reload can’t help you with that because changes are not actually written to disk
- But now you need to revert. Why not just undo the changes, the same way you made them?

**Comment by jpmonettas** [comment]
> But now you need to revert. Why not just undo the changes, the same way you made them?

You could have changed many things, like when you add print statements everywhere, or you are changing some library code in many places to see if a new feature make sense. You may have killed some of the buffers where you made the changes. So the only option is to restart the repl. It would be great for sanity and not restarting the repl to have something like reload `some-lib*` with whatever it takes to make that right (dependents and everything in topo order)

Does it make sense?

[FlowStorm](http://www.flow-storm.org/) could also use all of this, because when you set the compiler to instrument stuff you then need to recompile code. And there is no way currently to re-compile code (reload ns) by ns-regex, which clj-reload could provide.

**Comment by tonsky** [comment]
Okay, forgive me, I’m not trying to argue against it, just to understand the situation before making a decision. It will add a lot of complexity, so I want to be sure why it cannot be solved by other means.

So you changed namespaces without topological re-ordering, right?

1. Can’t you load them back with simple `(require ... :reload)` then?
2. What’s the minimal API that would work? Maybe `(require <regexp> :reload)`?
3. It still bothers me that when you chance namespace we can’t reload everything that depends on it (kindof what clj-reload is all about). It would be strange to support only half of that workflow

**Comment by jpmonettas** [comment]
Hey np, I understand that is not a trivial change, and maybe this wasn't even on the original scope of the project.

The simplest situation I can think of is, let's say you don't understand why a library is not working for you, so you start opening random functions and adding print statements or tweaking things to figure it out, adding some defs, atoms/reset! to capture values or whatever. After you figured it out you want to get rid of all the changes you made. The fastest is to have something like, re-load everything with this regex.

The problem with :reload and :reload-all is that it start with some namespace, so for :reload you need to know specific namespaces to reload. And to use :reload-all you have to know that all your changes were dependencies of that thing you are going to reload

**Comment by tonsky** [comment]
> so for :reload you need to know specific namespaces to reload

Yes but isn’t it the same with regex? Or do you plan to use `#".*"`?

**Comment by jpmonettas** [comment]
> Yes but isn’t it the same with regex? Or do you plan to use #".*"?

With regex for most cases doing `#"top-ns.*"` would be enough to be sure you reloaded an entire library without too much thinking.

I think it is a nice feature for a tool that focuses on reloading namespaces to have a way of reloading namespace that had been already loaded with a regex no matter what.

**Comment by velios** [comment]
@tonsky Is there any chance that adding this feature will be considered?

**Comment by tonsky** [comment]
We can try:

- If code doesn’t get much more complicated
- If performance in common case doesn’t suffer as a result
- If somebody provides a PR

I’m open to look at some code
